### DISCUSSION TITLE
Define commands on-the-fly without creating a class

### DISCUSSION THREAD

**Discussion by moodmosaic** [dubzzz/fast-check#3026] [Q&A] [answered]
The following interface should be implemented (by a class) in order to define a synchronous command:

```ts
interface Command<Model extends object, Real>
```

In addition to that, can a command be defined as a pair (or actually a *triple*) that holds the functions `check`, `run`, `toString`? Just trying to figure out if I can define commands without classes at all.

(Or, perhaps TypeScript has something similar to [F# Object Expressions](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/object-expressions) where you can implement an interface on-the-fly, without having to create a class.)

**Comment by dubzzz** [comment]
I think the on-the-fly should work fine with https://github.com/dubzzz/fast-check/blob/main/packages/fast-check/src/arbitrary/commands.ts#L40. The interface is mostly there to give the typing related contraints but not for extends needs (at least for the ICommand and derived stuffs).

**Reply by dubzzz** [reply]
But you'll probably need to setup of the fields requested by the contract.

**Comment by nth-commit** [comment, marked as answer]
Can confirm, I use commands without a class:

```typescript
type Command = fc.AsyncCommand<Model, Real>

const createCommand = (): Command => ({
  check: (m) => true,
  // ... etc.
})
```

**Reply by moodmosaic** [reply]
Awesome, thank you Michael.

**Reply by moodmosaic** [reply]
*Sidenote:* Great work on GalaxyCheck! I like the idea of [Sample](https://github.com/nth-commit/GalaxyCheck#transparency) as an attribute, among other stuff 🚀 (Also, greetings from the core Hedgehog team 🦔)

**Reply by nth-commit** [reply]
Thanks @moodmosaic! I definitely know who you are (we might have even interacted in the Hedgehog issues a long time ago), and I think Hedgehog is really great.

I need to go and brush up the docs and release a stable version. I thought having a C# day job again would bring me closer to it, but turns out that writing C# all day is more than enough C# 😅

**Reply by nth-commit** [reply]
Here's some examples if it helps:

I'm using a bespoke wrapper around the creating commands which lets me check some invariants after each command, have a default the `check` function, etc: https://github.com/nth-commit/xfiat/blob/620566179cf64cd3ab8b409145f838916804f49b/test/helpers/AsyncCommandFactory.ts

This is where I'm using that helper to specify some commands: https://github.com/nth-commit/xfiat/blob/620566179cf64cd3ab8b409145f838916804f49b/test/ReserveFunding/LiquidityAccumulatingModel.ts#L61-L101

And here is where I'm specifying my invariants: https://github.com/nth-commit/xfiat/blob/620566179cf64cd3ab8b409145f838916804f49b/test/ReserveFunding/LiquidityAccumulatingModel.ts#L31-L47
